core/src/main/scala/org/bykn/bosatsu/TypedExpr.scala

```diff
@@ -310,23 +320,65 @@ object TypedExpr {
       .distinct
   }
 
+  private def replaceVarType[A](te: TypedExpr[A], name: Identifier, tpe: Type): TypedExpr[A] = {
+    def recur(t: TypedExpr[A]) = replaceVarType(t, name, tpe)
+
+    te match {
+      case Generic(tv, in, tag) => Generic(tv, recur(in), tag)
+      case Annotation(term, tpe, tag) => Annotation(recur(term), tpe, tag)
+      case AnnotatedLambda(b, tpe, expr, tag) =>
+        // this is a kind of let:
+        if (b == name) {
+          // we are shadowing, so we are done:
+          te
+        }
+        else {
+          // no shadow
+          AnnotatedLambda(b, tpe, recur(expr), tag)
+        }
+      case Var(None, nm, _, tag) if nm == name => Var(None, name, tpe, tag)
+      case v@Var(_, _, _, _) => v
+      case App(fnT, arg, tpe, tag) =>
+        App(recur(fnT), recur(arg), tpe, tag)
+      case Let(b, e, in, r, t) =>
+        if (b == name) te // shadow
```

Comment:
this isn't quite right... if `r` is not recursive, we should unconditionally recurse on `e`.